Mage.Sets/src/mage/cards/f/FirestormPhoenix.java

```diff
@@ -0,0 +1,127 @@
+import mage.abilities.effects.Effect;
+import mage.abilities.effects.ReplacementEffectImpl;
+import mage.abilities.effects.common.ChooseACardNameEffect;
+import mage.abilities.effects.common.continuous.GainAbilityTargetEffect;
+import mage.abilities.effects.common.continuous.PlayWithHandRevealedEffect;
+import mage.abilities.keyword.FlyingAbility;
+import mage.abilities.keyword.ForetellAbility;
+import mage.abilities.keyword.HasteAbility;
+import mage.abilities.keyword.PlotAbility;
+import mage.cards.Card;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.cards.Cards;
+import mage.cards.CardsImpl;
+import mage.constants.CardType;
+import mage.constants.Duration;
+
+/**
+ *
+ * @author RobertFrosty
+ */
+
+public final class FirestormPhoenix extends CardImpl {
+
+    public FirestormPhoenix(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{4}{R}{R}");
+        
+        this.subtype.add(SubType.PHOENIX);
+        this.power = new MageInt(3);
+        this.toughness = new MageInt(2);
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // If FirestormPhoenix would die, return FirestormPhoenix to its owner's hand instead. Until that player's next turn, that player plays with that card revealed in their hand and can't play it.
+        Ability combAbility = new SimpleStaticAbility(new FirestormPhoenixEffect());
+        //combAbility.addEffect(new FirestormPhoenixEffect2());
+        this.addAbility(combAbility);
+    }
+
+    private FirestormPhoenix(final FirestormPhoenix card) {
+        super(card);
+    }
+
+    @Override
+    public FirestormPhoenix copy() {
+        return new FirestormPhoenix(this);
+    }
+}
+
+class FirestormPhoenixEffect extends ReplacementEffectImpl {
+	
+	FirestormPhoenixEffect() {
+		super(Duration.WhileOnBattlefield, Outcome.ReturnToHand);
+		staticText = "If Firestorm Phoenix would die, return Firestorm Phoenix to its owner's hand instead. Until that player's next turn, that player plays with that card revealed in their hand and can't play it.";
+	}
+	
+	private FirestormPhoenixEffect(final FirestormPhoenixEffect effect) {
+		super(effect);
+	}
+	
+	@Override
+	public FirestormPhoenixEffect copy() {
+		return new FirestormPhoenixEffect(this);
+	}
+	
+	//First attempt to write return to hand effect
+	@Override
+	public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+	    Player controller = game.getPlayer(source.getControllerId());
+	    if (controller != null) {
+	        Permanent permanent = ((ZoneChangeEvent) event).getTarget();
+	        Cards revealedCards = new CardsImpl();
+	        revealedCards.add(game.getCard(((ZoneChangeEvent) event).getTargetId()));
+	        if (permanent != null) {
+	            controller.moveCards(permanent, Zone.HAND, source, game);
+	            game.addEffect(new PlayWithHandRevealedEffect(TargetController.YOU), source);
```

Comment:
I think the reason this isn't working is because it's `Duration.WhileOnBattlefield` (you can see this in the constructor) and the source permanent isn't on the battlefield anymore.

Need custom effect like it that reveals the card if it's still in the hand, and discards itself otherwise. Probably needs custom `isInactive` to get the duration right too

Reply:
That's wrong effect. `PlayWithHandRevealedEffect` will reveal full hand instead single card. Commented `controller.revealCards` is correct. Also you must add additional game's effect here like `ExperimentalFrenzyRestrictionEffect` with fixed target (e.g. save `MageObjectReference mor` in effect's param) to restrict card play until next turn.

Reply:
Understood

Reply:
Hey, yeah, I know the effect is wrong; I couldn't get the custom effect that I wrote to fire, so I was just using a pre-built effect function so I could narrow down specifically what the issue was. I'll make sure to clean everything up before I do another PR! Appreciate the advice and feedback; I'll check out the ExperimentalFrenzyRestrictionEffect today! 

I appreciate all the information! I'll see if I can get something working, with this extra information.

Reply:
> I think the reason this isn't working is because it's `Duration.WhileOnBattlefield` (you can see this in the constructor) and the source permanent isn't on the battlefield anymore.
> 
> Need custom effect like it that reveals the card if it's still in the hand, and discards itself otherwise. Probably needs custom `isInactive` to get the duration right too

That makes a lot of sense; just took a look at the PlayWithHandRevealedEffect and I see what you mean. I also tried with a few other Duration values but couldn't get it to work with anything pre-built, so I agree, I definitely think it will need to be something custom; I see that Alhammarret, High Arbiter has a custom duration, and a custom isInactive as well, so I'll look deeper into that too. Again, really appreciate all the advice, learning a ton!

Reply:
Hey, sorry, it's been a busy week, so I'm just getting back to this now. 

> I think the reason this isn't working is because it's `Duration.WhileOnBattlefield` (you can see this in the constructor) and the source permanent isn't on the battlefield anymore.

You're definitely right; after some testing, the effect that gets added on line 102, game.addEffect(effect, source), seems to be discarded when the source leaves the battlefield, (when I return to hand). If the source doesn't change zones then everything works as expected.  I'm guessing there's some clean up func or event, that is removing all game effects that come from a source, when the source changes zones. 

My question is, what's the best way around this? The purpose of the custom isInactive would be to know when to discard the effect (I think), but how do I keep the game effect added inside of the replacement event from being removed when the source permanent changes zones?

Reply:
Zone.ALL for the ability

Reply:
Use duration.custom - it will keep effect forever. Plus override isInactive method or discard it by another way, e.g. on event check or apply.

Reply:
Thank you! I looked at game.java, effect.java, ability.java etc. but I didn't check the SimpleStaticAbility.java, silly of me. Really appreciate the help! That was the last piece of the puzzle, I've tested everything else, so I just need to put it all together; should be done shortly!

Reply:
Genuinely, thank you both so much for all the help! I just had to set the duration manually on the SimpleStaticAbility and that solved all my problems! <3 Hopefully I can finish this by the end of the day, and then I'll take a look at other cards / problems on the site.